fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -1017,6 +1022,18 @@ public void setVersion(int version) {
         this.version = version;
     }
 
+    /**
+     * Enable counter-based subspace key mechanism for indexes. If enabled, indexes subspace
+     * keys will be set based on a counter instead of index name.
+     */
+    public void enableIndexCounterBasedSubspaceKey() {
```

Comment:
I'm also not sure if this should return `void` or `this` (the latter being more idiomatic to `Builder`s). I had a similar problem when I made #381, where I needed to make a similar choice about setting the `evolutionValidator`.

Reply:
> having `enableIndexCounterBasedSubspaceKey` over `set` is intentional. It simplifies the logic.

How so?

> Do you have a use-case in mind for turning this off?

Something like the user initially wanted to use implicit subspace keys, but then they decided later that instead, they wanted to use more meaningful subspace keys because they cared more about looking at a key and knowing what it meant than having shorter keys.

Another use case where the user wants to set counter-based subspace keys on some record stores but not others. (Let's say on new record stores but not existing ones.) Then they can do something like `.setIndexCounterBasedSubspaceKeys(useCounter())` rather than `if (useCounter()) { .enable() }`.

> > I'm also not sure if this should return `void` or `this` (the latter being more idiomatic to `Builder`s).
> 
> Yeah, I think I have no preference on this. I thought void would be more suitable as I'm enabling it. If we decide to make it `set` I will change it `Builder`.

The more I think about it, I think it should return this. In particular, because you want to be able to do something like:

```java
RecordMetaDataBuilder builder = RecordMetaData.newBuilder()
     .enableIndexCounterBasedSubspaceKeys()
     .setRecords(MyRecordsDescriptor.getDescriptor());
```

to get counter-based keys for proto-defined indexes. (This assumes the extension options are processed in a predictable order, which ¯\\\_(ツ)_/¯.)

Reply:
> > having `enableIndexCounterBasedSubspaceKey` over `set` is intentional. It simplifies the logic.
> 
> How so?

It would disallow turning it off. I'm fine with making it setX too, I just don't like the idea of being able to turn it off/on back and forth, especially if the intent is to make this approach the default way (instead of name). It's similar to the case of the existing clients that are currently using name, and we don't want to mistakenly set their subspace keys to counter-based keys.

> Something like the user initially wanted to use implicit subspace keys, but then they decided later that instead, they wanted to use more meaningful subspace keys because they cared more about looking at a key and knowing what it meant than having shorter keys.

I think we currently have `index.setSubspaceKey` that takes care of this, if users really want this. 

> Another use case where the user wants to set counter-based subspace keys on some record stores but not others. (Let's say on new record stores but not existing ones.) Then they can do something like `.setIndexCounterBasedSubspaceKeys(useCounter())` rather than `if (useCounter()) { .enable() }`.
> 

Yeah, this I think is so preferential. I'm fine either way. As said before, my only reservation is the ability of turning off this feature.

> > > I'm also not sure if this should return `void` or `this` (the latter being more idiomatic to `Builder`s).
> > 
> > 
> > Yeah, I think I have no preference on this. I thought void would be more suitable as I'm enabling it. If we decide to make it `set` I will change it `Builder`.
> 
> The more I think about it, I think it should return this.

Sure, that's fine. I'll return `this`.

Reply:
Oh, I see. Turning it off is deeply problematic with the way it's currently implemented (as it assigns all keys at the end). As discussed in the other comment, I don't think it should assign all the prefixes there (because it makes using a `RecordMetaDataBuilder` without a meta-data proto unstable).

Reply:
OK, I went with this: I do not allow turning on / off, and in fact, I'm making it even stricter. One can enable this scheme only before calling `setRecords`. This would simplify reasoning about its consequences; i.e., if records in a `recordsDescriptor` already have indexes, we know how to deal with their subspace keys ahead of time. I see there are two main drawbacks, but I don't think they are terrible: 
- Clients should be explicit if they want to use `name`; i.e., they should call `setSubspaceKey`.
- if enabling subspace keys is controlled by some config / flag, they need to do: 
```
if(useCounter()) {
  builder.enableSubspaceKey();
}
```
I understand it may not be super convenient, but it suggests that turning on/off back and forth is not allowed. 


Also note that, existing clients that do not call `enableCounterBasedSubspaceKey` will continue using `name` for their subspace keys.

Reply:
I can live with this change, I suppose. It means that there isn't an upgrade path for existing clients who want to transition from using an existing meta-data definition with names to a counter-based assignment mechanism. I guess the user can themselves implement a counter and use it for new indexes...but that's a little sad that they are duplicating the work that's done here.

Reply:
Actually, I think there is an upgrade path: to start using a proto-defined meta-data. I think we want to get users on to that train anyway long term, so I think I can live with this, then.